docs/modules/storage/pages/high-density-memory.adoc

```diff
@@ -70,21 +70,28 @@ swap space left.
-*** Memory allocation and deallocation operations (except the ones
-requiring larger sizes than the page size) do not interact with the
-operating system mostly.
-*** For memory allocation, it tries to find the requested memory size inside
-the internal memory pools. If it cannot be found, then it interacts with the operating system.
+*** The allocator never frees memory blocks back to the operating system.
+Instead, it marks disposed memory blocks as available for reuse, so future allocations can use them again. Because of this,
+memory allocation and deallocation operations (except the ones
+requiring larger sizes than the page size) mostly do not interact with the
+operating system.
+*** When memory is needed, the allocator first tries to find a suitable free block inside its internal pools. If it cannot be found, then it interacts with the operating system to request a new page.
 *** Please note that the buddy memory allocation system may cause internal fragmentation depending on the size of the memory allocation request and configured page sizes. Since memory is allocated in block sizes that are a power-of-two, a request that does not exactly match the available block size will result in allocation of the next largest block, potentially leaving unused space. For example, when using 128 KB page sizes, a 68 KB request will result in the allocation of a 128 KB block, wasting 60 KB.
-*** Although the buddy memory allocation system reduces external fragmentation by merging contiguous and free buddy blocks of the same size into larger ones, it may still lead to external fragmentation if the system does not have sufficiently large free buddy blocks to allocate for the request. As a result, larger memory requests might be rejected, even if the total free memory is sufficient.
+*** The buddy memory allocation system reduces external fragmentation by merging contiguous free buddy blocks of the same size into larger ones. However, it may still lead to external fragmentation if the system does not have large enough free blocks to allocate for the request. As a result, larger memory requests might fail, even if the total free memory is enough.
+Because pages cannot be shared between threads, external fragmentation is more likely when:
+**** The total native memory is small.
+**** There is not enough memory available for each partition thread (2 GB total native memory and 8 partition threads means each thread will have ~256 MB).
+**** The workload includes frequent put/remove operations or a mix of small and large entries.
+*** Even if only a small amount of the memory is used, one partition thread may still run out of memory because it cannot borrow free blocks from others, and no free page is available from the operating system. As a result, a member can fail with NativeOutOfMemoryError even when total free memory appears large. This happens when a single thread's pool runs out of contiguous free blocks for the requested allocation size.
```

Comment:
> from the operating system.

more likely is hitting the configured `NativeMemoryConfig.capacity` than physically exhausting operating system memory (we might get killed by OOM killer in such case)